src/rez/package_maker__.py

```diff
@@ -31,7 +31,14 @@
         Or(basestring, [basestring]),
         {
             "command": Or(basestring, [basestring]),
-            Optional("requires"): [package_request_schema]
+            Optional("requires"): [package_request_schema],
+            Optional("run_on"): Or(basestring, [basestring]),
+            Optional("on_variants"): Or(
+                bool,
+                {
+                    "requires": [package_request_schema]
```

Comment:
"on_variants" is the scope that you would consider as the filter, whereas the fields within ("requires") can be thought of as the _type_ of filter. So this reads like "select on variants that require X". For eg, I haven't added it here (and probably won't) but in theory it would make sense to also have an "index" filter option, within the on_variants dict. And "on_variants" can't be named "filter" because that's too generic. We could go for "variants_filter", but is that any better than "on_variants"?

Example of use: on_variants.requires=["maya"] will run only on variants that (directly) require maya; on_variants.requires=["!nuke"] would run on all non-nuke variants.

I suppose you could have multiple 'requires' filters that would OR together, but I'm not sure that would ever be useful, and it would get quite confusing with conflict ops present. Ie, what would "!maya OR !nuke: mean? That would be the same as no requires filter at all. ANDed options aren't necessary, because you only need a single 'requires' field to describe that (eg ["maya", "!python-3"] will select any maya variant not using python-3).

Reply:
Note that my comment was not on the `on_variant` but on the `requires` key (from your comment I have the feeling you thought I was talking about the on_variant name...)

What about:

```
"on_variant": {
    "filter_type" (Or just type): "requirements", (eventually could be index too)
    "filter": ["maya"]
}
```
I feel this would make the intention clearer here than having yet another property called `requires`.

Reply:
I think this is ok, and if different filter types are added, this will avoid some confusion.

I would go with type/value though rather than filter_type/filter (the on_variant dict _is_ the filter)

Reply:
Great. I'm fine with `type` and `value`.